### PR TITLE
Enter/leave chat messages

### PR DESCRIPTION
**Pull request by o01eg**
Adds support to show in chat "Player entered game" and "Player left game" messages.

### CODE REVIEW

**Comment by geoffthemedio** [comment]
This should be implemented as a network message that the client responds to by generating the message text in chat locally, not as a server-side composed text message that will be the same regardless of players' local language settings.

**Comment by o01eg** [comment]
> This should be implemented as a network message that the client responds to by generating the message text in chat locally,

I want those messages could be stored.
> not as a server-side composed text message that will be the same regardless of players' local language settings

It translates on the client depending on client's string table file.

File: server/ServerFSM.cpp [hunk 1]
```diff
@@ -456,6 +468,19 @@ bool ServerFSM::EstablishPlayer(const PlayerConnectionPtr& player_connection,
             client_type == Networking::CLIENT_TYPE_HUMAN_OBSERVER ||
             client_type == Networking::CLIENT_TYPE_HUMAN_PLAYER)
         {
+            // add "player enter game" message
+            boost::posix_time::ptime timestamp = boost::posix_time::second_clock::universal_time();
+            std::string data = player_name + " [[" + UserStringNop("PLAYER_ENTERED_GAME") + "]]";
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: It translates `PLAYER_ENTERED_GAME` when the chatbox sees the `[[` and `]]`, but this is doing the string composition in code with `+` instead of substitution with `FlexibleFormat`, which would allow the stringtable / translation to control where and how the name appears in the string.

Comment (o01eg) [reply server/ServerFSM.cpp]: Changed to boost::format

File: server/ServerFSM.cpp [hunk 2]
```diff
@@ -274,7 +274,7 @@ void ServerFSM::HandleNonLobbyDisconnection(const Disconnection& d) {
         }
         // add "player left game" message
         boost::posix_time::ptime timestamp = boost::posix_time::second_clock::universal_time();
-        std::string data = player_connection->PlayerName() + " [[" + UserStringNop("PLAYER_LEFT_GAME") + "]]";
+        std::string data = (boost::format("%s [[PLAYER_LEFT_GAME]]") % player_connection->PlayerName()).str();
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: This is no different from using `+`.

The point is that the string composition should be done client-side so the stringtable of the client can determine how and where to put the parameter into the string.

File: server/ServerFSM.cpp (hunk 1 above)
Comment (geoffthemedio) [reply server/ServerFSM.cpp]: You missed the part about using the stringtable entry to determine where and how the name appears in the string. Just using `boost::format` instead of `+` but keeping the position of the name in the string the same is worse than it was before.

File: server/ServerFSM.cpp (hunk 2 above)
Comment (geoffthemedio) [reply server/ServerFSM.cpp]: Not all languages have Subject-Verb-Object ordering, so you shouldn't assume the player name will appear first in the string. If the substitution is done on the client, and has a playername parameter, the stringtable can specify where the substituted name should appear in the displayed text.

Comment (o01eg) [reply server/ServerFSM.cpp]: Does `[[STRINGTABLE_ENTRY]]` have arguments support so it could be written like `[[PLAYER_LEFT_GAME,Player name]]`?

Comment (geoffthemedio) [reply server/ServerFSM.cpp]: No, see here:

https://github.com/freeorion/freeorion/blob/master/UI/ChatWnd.cpp#L37

It's a bit hard to follow, but it scans through the input chat message, copying everything to the filtered message until it sees `[[`, then it searches for a stringtable entry that matches what comes next before the following `]]`. It then copies the stringtable entry's value instead of the `[[` and `]]` and the stuff between them.

No format call occurs, so no formatting gets done like for `VarText`, and no recursive macro substitution is done, like when loading the stringtable or scripts and processing macros.

Comment (o01eg) [reply server/ServerFSM.cpp]: Ok, I'll postpone this PR until some stringtable arguments would be implemented.

Comment (geoffthemedio) [reply server/ServerFSM.cpp]: https://github.com/freeorion/freeorion/commit/783fa637b5733a2730aa69e18ff441daa4aaad1c should support client-side string or stringtable entry substitutions.

Comment (o01eg) [reply server/ServerFSM.cpp]: Thanks, updated PR according to this change.

**Comment by o01eg** [comment]
Rebased. Used new argument support for enter/leave messages.

File: server/ServerFSM.cpp [hunk 3]
```diff
@@ -285,6 +285,18 @@ void ServerFSM::HandleNonLobbyDisconnection(const Disconnection& d) {
                 m_server.DropPlayerEmpireLink(id);
             }
         }
+        // add "player left game" message
+        boost::posix_time::ptime timestamp = boost::posix_time::second_clock::universal_time();
+        std::string data = (boost::format("[[PLAYER_LEFT_GAME,%s]]") % player_connection->PlayerName()).str();
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: `boost::format` isn't necessary here. It'd be fine to do just:
`std::string data = "[[PLAYER_LEFT_GAME," + player_connection->PlayerName()).str() + "]]";`

That said, it's probably slightly easier to read the current version, so OK.

Comment (o01eg) [reply server/ServerFSM.cpp]: Fixed.

File: server/ServerFSM.cpp [hunk 4]
```diff
@@ -285,6 +285,18 @@ void ServerFSM::HandleNonLobbyDisconnection(const Disconnection& d) {
                 m_server.DropPlayerEmpireLink(id);
             }
         }
+        // add "player left game" message
+        boost::posix_time::ptime timestamp = boost::posix_time::second_clock::universal_time();
+        std::string data = "[[PLAYER_LEFT_GAME," + player_connection->PlayerName() + "]]";
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: arguably the `PLAYER_LEFT_GAME` should be wrapped in `UserStringNoOp` ...

Comment (o01eg) [reply server/ServerFSM.cpp]: Fixed

File: server/ServerFSM.cpp [hunk 5]
```diff
@@ -504,7 +504,7 @@ bool ServerFSM::EstablishPlayer(const PlayerConnectionPtr& player_connection,
         {
             // add "player enter game" message
             boost::posix_time::ptime timestamp = boost::posix_time::second_clock::universal_time();
-            std::string data = "[[PLAYER_ENTERED_GAME," + player_connection->PlayerName() + "]]";
+            std::string data = std::string("[[") + UserStringNop("PLAYER_ENTERED_GAME") + "," + player_connection->PlayerName() + "]]";
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp:507]: Is the `std::string` wrapping `"[["` neccessary?

Comment (o01eg) [reply server/ServerFSM.cpp:507]: Yes, without it I get `const char* + const char*` error.
